RFC-0015-Tracer-module-scope-extension.md

```diff
@@ -0,0 +1,74 @@
+## Summary
+Extending `TracingState.push_scope` and `TracingState.pop_scope` with function overload to take module inputs, attributes, and outputs as additional arguments. Enalbing more information of the module call to be recorded when `recording_scopes` is turned on.
+
+## Motivation
+A particular `Scope` object is created for each `nn.Module` `forward` call when `recording_scopes` is turned on. Each node within the `forward` call is annotated with that `Scope` object. This enables ONNX export to group nodes based on scope of original `nn.Module`.
+
+However, with current approach, the information of inputs, outputs, and attributes of `nn.Module` call is missing. Graph analysis can be done to figure out which `Value` should be input/output, based on the scope location of their creation and use. Yet it is no longer possible to retrieve the input/output order or input/output name. For module attributes, it is even worse as some attributes are only used in module initialization, and not visible/retrievable in traced graphs.
+
+The goals are
+
+* Goal: enable jit (ONNX) passes to retrieve inputs/outputs/attributes information for each `Scope`.
+
+* Goal: only affect code path of ONNX export when this feature is specifically turned on.
+
+## User description
+
+Users are not aware of this implementation detail.
+
+## Implementation description
+
+**TracingState.**
+```c++
+  py::class_<TracingState, std::shared_ptr<TracingState>>(
+      m, "TracingState", py::dynamic_attr())
+      .def(
+          "push_scope",
+          [](TracingState& s,
+             const std::string& scope_name,
+             const py::tuple& input_tuple,
+             const py::dict& input_dict,
```

Comment:
(1) correct. *args and **kwargs
(2) likely not, since args is tuple.
(3) tracing cannot handle optional input. We observe what's provided in sample input for trace call.